hail/src/main/scala/is/hail/expr/ir/functions/LocusFunctions.scala

```diff
@@ -349,7 +370,12 @@ object LocusFunctions extends RegistryFunctions {
         )
     }
 
-    registerCodeWithMissingness("LocusInterval", TString, TInt32, TInt32, TBoolean, TBoolean, TBoolean, tinterval("T"), null) {
+    registerCodeWithTypeAndMissingness("LocusInterval", TString, TInt32, TInt32, TBoolean, TBoolean, TBoolean, tinterval("T"), {
+      (returnType: Type, _: PType, _: PType, _: PType, _: PType, _: PType, _: PType) => {
+        val lPTyp = returnType.asInstanceOf[TInterval].pointType.asInstanceOf[TLocus]
+        PCanonicalInterval(PCanonicalLocus(lPTyp.rg))
```

Comment:
locus here can't be required. I don't fully understand why, as the emit code returned from this function sets the returned value to missing if any argument is missing, or if the locus is null (so I should never be able to get an interval with a null locus, or so I thought)

Reply:
Ah, this is because of invalid_missing. Interesting.

```python
def locus_interval(contig,
                   start,
                   end,
                   includes_start=True,
                   includes_end=False,
                   reference_genome: Union[str, ReferenceGenome] = 'default',
                   invalid_missing=False) -> IntervalExpression:
```

```scala
def toLocusInterval(...) = { 
   ... 
    if (!isValidLocus(start.contig, if (includesStart) start.position else start.position + 1)) {
      if (invalidMissing)
        return null
```

So we can get a null value even though the locus is required to be provided

Reply:
just look at the generated EmitCode -- there's an OR case that doesn't look at the missingness of the arguments, so it's possible to get a missing value with all arguments required. Return type always needs to be optional.

Reply:
Don't look at the python interface, look at the `m` argument in the EmitCode returned. That's the important kernel that tells you *exactly* what the missingness behavior of this function is.

Reply:
Does it not help to understand why the last argument can be isNull?

Reply:
The point I'm trying to make is that while you can look at the vertical slice of a specific function to understand everything about it, this isn't necessary and it will both slow you down and make it more likely that you won't be as focused on the specific change that motivates the work. The way to work fast is by keeping the process of making a change simple and small.

Reply:
you don't need to understand *why* the last condition has an `isNull`, you only need to see that it's there to note that there's a missing case when all arguments are present

Reply:
> The way to work fast is by keeping the process of making a change simple and small.

I think it should be obvious that I looked at the isNull before anything else, I don't understand why you would assume that I looked at some hard to find function before looking at the one in front of me.

The issue wasn't that I was mistakenly setting outer requiredeness (on PInterval), which is what the presence of the local isNull would help me avoid. That was obvious. The issue was that even with PInterval being non-required, PLocus couldn't be required. I didn't understand this because when the Interval is not `isNull`, every other argument must be present, and therefore the PLocus can be required. In order to understand why this wasn't the case, the only thing I could do, besides stare at this registerCodeWithMissingness body some more, was look at the invocation that led to the issue.

Reply:
> I think it should be obvious that I looked at the isNull before anything else, I don't understand why you would assume that I looked at some hard to find function before looking at the one in front of me. Could you explain?

I apologize for the unmerited critique, and I recognize that you definitely arrived to the correct conclusion here! I've been missing the point about the *locus* as well as the *interval*. To determine whether the locus can be required, you definitely do need to look at `makeInterval` and therefore toLocusInterval, which certainly isn't something you'd naively expect to modify missingness. I'm sorry about that! This is quite subtle.

Reply:
No worries!

Reply:
Just wanted to make sure that I explained my thought process, and if it was wrong, remedy :)